doc/rados/operations/crush-map-edits.rst

```diff
@@ -525,6 +540,16 @@ A rule takes the following form::
                  final CRUSH mapping transformation is therefore 1, 2, 3, 4, 5
                  → 1, 2, 6, 4, 5.
 
+``step choosemsr {num} type {bucket-type}``
+   :Description: Selects a num buckets of type bucket-type.  msr_firstn and msr_indep
+		 must use choosemsr rather than choose or chooseleaf.
```

Comment:
IIUC, `choose` has a bit of disadvantage against `chooseleaf` IF in case the host has most(or maybe all!) the OSDs down but since this is MSR, does this mean we attain data distribution balance same as `chooseleaf`?

Reply:
I'm not sure exactly what you mean, the behavior of chooseleaf and choose differs by more than data placement.  I think you are referring to the difference between

```
...
step choose 3 type host
step choose 1 type osd
...
```

and

```
...
step chooseleaf 3 type host
...
```

In the former case, marking an osd down will result in another osd from the same host being chosen.  As more osds are marked out, the remaining OSDs on the same host will be overloaded until eventually it becomes a placement failure if there are no remaining in osds.

In the latter case, when we hit a down osd within the chooseleaf, we'll retry a different host resulting in a more even distribution as PGs will leave the host with down OSDs proportionately to how many OSDs are down.

```
...
step choosemsr 3 type host
step choosemsr 1 type osd
...
```

would retry the full descent when it hits an out OSD resulting in a redistribution as with the chooseleaf example above.  That's essentially the point -- msr rules sort of generalize chooseleaf's behavior and make it unnecessary.